### PR TITLE
Normalize obligation and expected trait_refs in confirm_poly_trait_refs

### PR DESCRIPTION
**Pull request by compiler-errors**
Consolidate normalization the obligation and expected trait refs in `confirm_poly_trait_refs`. Also, _always_ normalize these trait refs -- we were already normalizing the obligation trait ref when confirming closure and generator candidates, but this does it for fn pointer confirmation as well. 

This presumably does more work in the case that the obligation's trait ref is already normalized, but we can see from the perf runs in #94070, it actually (paradoxically, perhaps) improves performance when paired with logic that normalizes projections in fulfillment loop.

### CODE REVIEW

**Comment by compiler-errors** [comment]
Forgot to remove a deleted test stderr..

**Comment by jackh726** [comment]
I think we can land this separately, but I need to review later.

**Comment by compiler-errors** [comment]
Cool, I will make it not a draft and give it a real title and description, then.

File: src/test/ui/generic-associated-types/bugs/issue-88382.stderr [hunk 1]
```diff
@@ -2,18 +2,18 @@ error[E0631]: type mismatch in function arguments
   --> $DIR/issue-88382.rs:27:40
    |
 LL |     do_something(SomeImplementation(), test);
-   |     ------------                       ^^^^ expected signature of `for<'a> fn(&mut <SomeImplementation as Iterable>::Iterator<'a>) -> _`
+   |     ------------                       ^^^^ expected signature of `for<'r> fn(&'r mut std::iter::Empty<usize>) -> _`
```
Comment (jackh726) [inline src/test/ui/generic-associated-types/bugs/issue-88382.stderr:5]: This test highlights that something wonky is going on. It changes were we aren't able to normalize.

Comment (compiler-errors) [reply src/test/ui/generic-associated-types/bugs/issue-88382.stderr:5]: Yeah, this is very interesting. Possibly due to changes in the ordering of resolving the inference variable in `&mut <_ as Iterable>::Iterator<'a>` and normalizing it. Now that we're eagerly normalizing the trait ref, we probably are substituting that value before we have a chance to unify it with something else, possibly another projection type with an inference variable.

**Comment by CAD97** [comment]
Just FYI: this likely changed inference to (correctly) accept more cases: https://www.reddit.com/r/rust/comments/t8ifzy/type_inference_fails_on_stable_but_succeeds_on/?utm_medium=android_app&utm_source=share

**Comment by compiler-errors** [comment]
@CAD97 this seems consistent with the changes in this this PR (and the issues I was originally focusing on when I first wrote this code in #93361). Thanks very much for noting this!

**Pull request merged**